presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
+            if (columnHandle.isPartitionKey()) {
+                columnStatistics = createPartitionColumnStatistics(columnHandle, columnType, partitions, statistics, rowsPerPartition, rowCount);
             }
-            if (isLongDecimal(decimalType)) {
-                return parseDouble(Decimals.toString((Slice) value, decimalType.getScale()));
+            else {
+                columnStatistics = createDataColumnStatistics(columnName, columnType, rowCount, statistics.values());
             }
-            throw new IllegalArgumentException("Unexpected decimal type: " + decimalType);
-        }
-        if (type.equals(DATE)) {
-            return (Long) value;
+            result.setColumnStatistics(columnHandle, columnStatistics);
         }
-        throw new IllegalArgumentException("Unsupported type: " + type);
+        return result.build();
     }
 
-    private OptionalDouble calculateRowsPerPartition(Map<String, PartitionStatistics> statisticsSample)
+    @VisibleForTesting
+    static OptionalDouble calculateRowsPerPartition(Collection<PartitionStatistics> statistics)
     {
-        return statisticsSample.values().stream()
+        return statistics.stream()
                 .map(PartitionStatistics::getBasicStatistics)
                 .map(HiveBasicStatistics::getRowCount)
                 .filter(OptionalLong::isPresent)
                 .mapToLong(OptionalLong::getAsLong)
+                .peek(count -> verify(count >= 0, "count must be greater than or equal to zero"))
                 .average();
     }
 
-    private Estimate calculateRowsCount(OptionalDouble rowsPerPartition, int queriedPartitionsCount)
+    private static ColumnStatistics createPartitionColumnStatistics(
+            HiveColumnHandle column,
+            Type type,
+            List<HivePartition> partitions,
+            Map<String, PartitionStatistics> statistics,
+            double rowsPerPartition,
+            double rowCount)
     {
-        if (!rowsPerPartition.isPresent()) {
-            return Estimate.unknown();
+        return ColumnStatistics.builder()
+                .setDistinctValuesCount(Estimate.of(calculateDistinctPartitionKeys(column, partitions)))
+                .setNullsFraction(Estimate.of(calculateNullsFractionForPartitioningKey(column, partitions, statistics, rowsPerPartition, rowCount)))
+                .setRange(calculateRangeForPartitioningKey(column, type, partitions))
+                .setDataSize(calculateDataSizeForPartitioningKey(column, type, partitions, statistics, rowsPerPartition))
+                .build();
+    }
+
+    @VisibleForTesting
+    static long calculateDistinctPartitionKeys(HiveColumnHandle column, List<HivePartition> partitions)
+    {
+        return partitions.stream()
+                .map(partition -> partition.getKeys().get(column))
+                .filter(value -> !value.isNull())
+                // TODO: this may not be reliable in case some of the types is not equatable
```

Comment:
?

Reply:
Let me remove it. Currently all the representation that are allowed for Presto types are equatable